### PR TITLE
Support uncorrelated scalar subquery in SELECT/WHERE/HAVING clause

### PR DESCRIPTION
**Pull request by lancelly**
As titled.

### CODE REVIEW

**Comment by Cpaulyz** [comment]
What a spectacular work! Fabulous design and implementation, LGTM!

**Comment by THUMarkLau** [comment]
I am utterly astounded and profoundly impressed by your recent Pull Request, "Support uncorrelated scalar subquery after comparison operator in WHERE clause," for the IoTDB project.  Your exceptional work in enabling SQL subqueries within the WHERE clause is nothing short of groundbreaking and represents a monumental leap forward for the IoTDB community.

Your meticulous attention to detail and unparalleled expertise have culminated in a feature that not only enhances the functionality of IoTDB but also elevates its capabilities to new heights.  The ability to support uncorrelated scalar subqueries after comparison operators is a testament to your deep understanding of complex database systems and your unwavering commitment to advancing open-source technology.

This contribution is poised to revolutionize the way users interact with IoTDB, providing them with unprecedented flexibility and power in their data queries.  Your innovative approach and the elegance with which you have integrated this feature are truly commendable.  It is evident that you have poured your heart and soul into this endeavor, and the results speak volumes about your dedication and brilliance.

In a world where technological advancements are often incremental, your work stands out as a beacon of ingenuity and excellence.  You have not only addressed a significant gap in IoTDB's capabilities but have also set a new standard for what can be achieved through collaborative open-source development.

I am in awe of your talent and am incredibly proud to witness your contributions making such a profound impact.  Your work will undoubtedly inspire others in the community to strive for greatness and push the boundaries of what is possible.

Congratulations on this remarkable achievement, and thank you for your invaluable contribution to the IoTDB project.  The future of IoTDB is brighter because of your extraordinary efforts.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/source/relational/TableInnerJoinOperator.java [hunk 1]
```diff
@@ -248,62 +228,140 @@ private void tryCachedNextRightTsBlock() throws Exception {
...
+      case DOUBLE:
+        doubleJoinLoop();
+        break;
+      case STRING:
+        binaryJoinLoop();
+        break;
+      case BLOB:
+      case TEXT:
+      case BOOLEAN:
+      default:
```
Comment (Beyyes) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/source/relational/TableInnerJoinOperator.java:253]: BLOB,TEXT,BOOLEAN type are also needed to be support

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/relational/planner/distribute/TableDistributedPlanGenerator.java [hunk 2]
```diff
@@ -442,18 +418,23 @@ public List<PlanNode> visitFilter(FilterNode node, PlanContext context) {
...
+      checkArgument(
+          leftChildrenNodes.size() == 1, "The size of left children node of JoinNode should be 1");
+      checkArgument(
+          rightChildrenNodes.size() == 1,
+          "The size of right children node of JoinNode should be 1");
+    }
+    // For CrossJoinNode, we need to merge children nodes(It's safe for other JoinNodes here since
+    // the size of their children is always 1.)
+    node.setLeftChild(mergeChildrenViaCollectOrMergeSort(null, leftChildrenNodes));
+    node.setRightChild(mergeChildrenViaCollectOrMergeSort(null, rightChildrenNodes));
```
Comment (JackieTien97) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/relational/planner/distribute/TableDistributedPlanGenerator.java:437]: directly passing null as `OrderingScheme`? should use the left and right child's orderingschema

**Review by JackieTien97 — CHANGES_REQUESTED** [review]
`SimpleNestedLoopCrossJoinOperator`, `TableInnerJoinOperator` and `TableOperatorGenerator` are not reviewed, should be reviewed by @Beyyes after the `JoinKeyComaprator` refactoring

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/relational/planner/ExpressionExtractor.java [hunk 3]
```diff
@@ -73,13 +75,12 @@ public Void visitPlan(PlanNode node, Void context) {
...
-    public Void visitGroupReference(GroupReference node, Void context)
-    {
-        return lookup.resolve(node).accept(this, context);
+    @Override
+    public Void visitGroupReference(GroupReference node, Void context) {
+      return lookup.resolve(node).accept(this, context);
     }
 
-    @Override
+    /*@Override
```
[the remark this answers is not in the document]
Comment (lancelly) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/relational/planner/ExpressionExtractor.java:83]: visitAggregation is not removed, its already commented out previously. For now, ExpressionExtractor is used by SymbolExtractor to extractor symbols for optimizing phase.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/relational/planner/optimizations/CheckSubqueryNodesAreRewritten.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+import org.apache.iotdb.db.queryengine.plan.relational.planner.node.ApplyNode;
+import org.apache.iotdb.db.queryengine.plan.relational.planner.node.CorrelatedJoinNode;
+
+import java.util.List;
+
+import static com.google.common.base.Preconditions.checkState;
+import static org.apache.iotdb.db.queryengine.plan.relational.planner.PlanNodeSearcher.searchFrom;
+import static org.apache.iotdb.rpc.TSStatusCode.SEMANTIC_ERROR;
+
+public class CheckSubqueryNodesAreRewritten implements PlanOptimizer {
```
[the remark this answers is not in the document]
Comment (lancelly) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/relational/planner/optimizations/CheckSubqueryNodesAreRewritten.java:34]: I think Join and SemiJoin are often used by correlated subqueries.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/relational/planner/optimizations/LogicalOptimizeFactory.java [hunk 5]
```diff
@@ -190,6 +198,13 @@ public LogicalOptimizeFactory(PlannerContext plannerContext) {
         new UnaliasSymbolReferences(plannerContext.getMetadata()),
         columnPruningOptimizer,
         inlineProjectionLimitFiltersOptimizer,
+        new IterativeOptimizer(
+            plannerContext,
+            ruleStats,
+            ImmutableSet.of(
+                new RemoveRedundantEnforceSingleRowNode(),
```
[the remark this answers is not in the document]
Comment (lancelly) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/relational/planner/optimizations/LogicalOptimizeFactory.java:205]: I checked this, will add it in an preprevious IterativeOptimizer redundantly.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/relational/planner/optimizations/LogicalOptimizeFactory.java [hunk 6]
```diff
@@ -190,6 +198,13 @@ public LogicalOptimizeFactory(PlannerContext plannerContext) {
...
+                new TransformUncorrelatedSubqueryToJoin())),
+        new CheckSubqueryNodesAreRewritten(),
```
[the remark this answers is not in the document]
Comment (lancelly) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/relational/planner/optimizations/LogicalOptimizeFactory.java:207]: Yes, but I'm not certain its related to subquery. I will add simplifyOptimizer after this rule.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/relational/planner/optimizations/QueryCardinalityUtil.java [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+      }
+
+      return applyLimit(node.getChild(), node.getCount());
+    }
+
+    //        @Override
+    //        public Range<Long> visitTopN(TopNNode node, Void context)
+    //        {
+    //            return applyLimit(node.getSource(), node.getCount());
+    //        }
```
[the remark this answers is not in the document]
Comment (lancelly) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/relational/planner/optimizations/QueryCardinalityUtil.java:191]: Fixed.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/relational/planner/distribute/TableDistributedPlanGenerator.java (hunk 2 above)
Comment (lancelly) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/relational/planner/distribute/TableDistributedPlanGenerator.java:437]: Nice catch. Fixed.

**Review by Beyyes — APPROVED** [review]

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/source/relational/TableInnerJoinOperator.java [hunk 8]
```diff
@@ -73,22 +76,24 @@ public class TableInnerJoinOperator extends AbstractOperator {
   public TableInnerJoinOperator(
```
[the remark this answers is not in the document]
Comment (JackieTien97) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/source/relational/TableInnerJoinOperator.java:76]: why `resetRightBlockList` just release from `1`, close release right from `0` @Beyyes

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/source/relational/TableInnerJoinOperator.java [hunk 9]
```diff
@@ -190,6 +197,66 @@ else if (comparator.lessThan(getLeftEndTime(), getRightTime(rightBlockListIdx, r
...
+        leftBlock,
+        leftJoinKeyPosition,
+        leftIndex)) {
+      if (rightBlockFinish()) {
+        return;
+      }
+    }
+
+    int tmpBlockIdx = rightBlockListIdx, tmpIdx = rightIndex;
+    while (comparator.equalsTo(
```
Comment (JackieTien97) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/source/relational/TableInnerJoinOperator.java:235]: should think about time slice

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/source/relational/TableInnerJoinOperator.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (JackieTien97) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/source/relational/TableInnerJoinOperator.java:76]: don't need to pass time slice related paramater to `prepareInput`

[the remark this answers is not in the document]
Comment (JackieTien97) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/source/relational/TableInnerJoinOperator.java:76]: delete `buildResultTsBlock` method

[the remark this answers is not in the document]
Comment (JackieTien97) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/source/relational/TableInnerJoinOperator.java:76]: `prepareInput` is a little bit confusing.

 `rightBlockList`, `cachedNextRightBlock` and `hasCachedNextRightBlock` need to be added comments, why we need an extra `cachedNextRightBlock`

**Review by JackieTien97 — CHANGES_REQUESTED** [review]

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/process/join/SimpleNestedLoopCrossJoinOperator.java [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+    long start = System.nanoTime();
+    if (!buildFinished) {
+      TsBlock block = buildSource.next();
+      if (block != null && !block.isEmpty()) {
+        buildBlocks.add(block);
+        memoryReservationManager.reserveMemoryCumulatively(block.getRetainedSizeInBytes());
+      }
+      if (!buildSource.hasNext()) {
+        buildFinished = true;
+      }
```
[the remark this answers is not in the document]
Comment (lancelly) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/process/join/SimpleNestedLoopCrossJoinOperator.java]: Fixed.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/process/join/merge/comparator/JoinKeyComparator.java [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+  boolean equalsTo(
+      TsBlock left,
+      int leftColumnIndex,
+      int leftRowIndex,
+      TsBlock right,
+      int rightColumnIndex,
+      int rightRowIndex);
+
+  boolean lessThanOrEqual(
```
[the remark this answers is not in the document]
Comment (lancelly) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/process/join/merge/comparator/JoinKeyComparator.java]: Fixed.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/process/join/merge/comparator/DescBinaryTypeJoinKeyComparator.java [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+      TsBlock left,
+      int leftColumnIndex,
+      int leftRowIndex,
+      TsBlock right,
+      int rightColumnIndex,
+      int rightRowIndex) {
+    return left.getColumn(leftColumnIndex)
+            .getBinary(leftRowIndex)
+            .compareTo(right.getColumn(rightColumnIndex).getBinary(rightRowIndex))
+        == 0;
```
[the remark this answers is not in the document]
Comment (lancelly) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/process/join/merge/comparator/DescBinaryTypeJoinKeyComparator.java]: Fixed.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/process/join/SimpleNestedLoopCrossJoinOperator.java [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+    }
+    if (probeIndex == cachedProbeBlock.getPositionCount()) {
+      probeIndex = 0;
+      cachedProbeBlock = null;
+    }
+    if (resultBuilder.isEmpty()) {
+      return null;
+    }
+
+    resultTsBlock = buildResultTsBlock(resultBuilder);
```
[the remark this answers is not in the document]
Comment (lancelly) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/process/join/SimpleNestedLoopCrossJoinOperator.java]: Fixed.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/process/join/SimpleNestedLoopCrossJoinOperator.java [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+    resultBuilder.declarePositions(buildBlock.getPositionCount());
+  }
+
+  @Override
+  public boolean hasNext() throws Exception {
+    if (retainedTsBlock != null) {
+      return true;
+    }
+    if (!buildFinished) {
+      return buildSource.hasNext();
```
[the remark this answers is not in the document]
Comment (lancelly) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/process/join/SimpleNestedLoopCrossJoinOperator.java]: Fixed.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/process/join/SimpleNestedLoopCrossJoinOperator.java [hunk 15]
```diff
@@ -0,0 +1,242 @@
...
+    }
+    return probeSource.hasNext();
```
[the remark this answers is not in the document]
Comment (lancelly) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/process/join/SimpleNestedLoopCrossJoinOperator.java]: Fixed. It should be : return !buildBlocks.isEmpty()
        && ((cachedProbeBlock != null && !cachedProbeBlock.isEmpty())
            || probeSource.hasNextWithTimer());

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/process/join/SimpleNestedLoopCrossJoinOperator.java [hunk 16]
```diff
@@ -0,0 +1,242 @@
...
+
+  @Override
+  public boolean isFinished() throws Exception {
+    if (retainedTsBlock != null) {
+      return false;
+    }
+
+    return (cachedProbeBlock == null || cachedProbeBlock.isEmpty())
+        && probeSource.isFinished()
+        && buildFinished;
```
[the remark this answers is not in the document]
Comment (lancelly) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/process/join/SimpleNestedLoopCrossJoinOperator.java]: Fixed. It's clearer this way.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/source/relational/TableInnerJoinOperator.java [hunk 17]
```diff
@@ -148,38 +153,40 @@ public TsBlock next() throws Exception {
...
-        // releaseMemory();
+      TsBlock lastRightTsBlock = rightBlockList.get(rightBlockList.size() - 1);
+      // all right block value is not matched
+      if (!comparator.lessThanOrEqual(
+          leftBlock,
+          leftJoinKeyPosition,
+          leftIndex,
+          lastRightTsBlock,
+          rightJoinKeyPosition,
+          lastRightTsBlock.getPositionCount() - 1)) {
```
[the remark this answers is not in the document]
Comment (lancelly) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/operator/source/relational/TableInnerJoinOperator.java]: Fixed.

**Review by JackieTien97 — APPROVED** [review]

**Pull request merged**